internal/provider/resource_instance.go

```diff
@@ -1681,3 +1720,56 @@ func (f instanceExternalIPValidator) ValidateSet(ctx context.Context, req valida
 		return
 	}
 }
+
+// ModifyPlanForHostnameDeprecation modifies the plan to support the
+// deprecation of `host_name` in favor of `hostname`. This must be
+// added to both the deprecated `host_name` attribute and the new
+// `hostname` attribute. This function is responsible for setting
+// [stringplanmodifier.RequiresReplaceIfFuncResponse.RequiresReplace] to `true`
+// when it's deemed the user is actually changing the hostname value in the
+// configuration or `false` when the user is just updating their configuration
+// to comply with the deprecation.
+func ModifyPlanForHostnameDeprecation(ctx context.Context, req planmodifier.StringRequest, resp *stringplanmodifier.RequiresReplaceIfFuncResponse) {
+	// Check which attribute this modifier function is being called on as the logic
+	// is vice versa for `host_name` and `hostname`.
+	switch attribute := req.Path.String(); attribute {
+	case "hostname":
+		var hostnameDeprecated types.String
+		diags := req.Config.GetAttribute(ctx, path.Root("host_name"), &hostnameDeprecated)
+		resp.Diagnostics.Append(diags...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+
+		// The deprecated `host_name` attribute has a value. We do not need to replace
+		// the resource just because the new `hostname` attribute doesn't have a value.
+		if !hostnameDeprecated.IsNull() && !hostnameDeprecated.IsUnknown() {
+			return
+		}
+	case "host_name":
+		var hostname types.String
+		diags := req.Plan.GetAttribute(ctx, path.Root("hostname"), &hostname)
+		resp.Diagnostics.Append(diags...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+
+		// The new `hostname` attribute has a value. We do not need to replace the
+		// resource just because the deprecated `host_name` attribute doesn't have
+		// a value.
+		if !hostname.IsUnknown() && !hostname.IsNull() {
+			return
+		}
+	default:
+		resp.Diagnostics.AddAttributeError(
+			req.Path,
+			fmt.Sprintf("Invalid attribute %s", attribute),
+			"HostnamePlanModifier can only be used for instance hostname.",
+		)
```

Comment:
```suggestion
		resp.Diagnostics.AddAttributeError(
			req.Path,
			fmt.Sprintf("Invalid plan modifier for attribute %s", attribute),
			"ModifyPlanForHostnameDeprecation can only be used for instance hostname attributes.",
		)
```

I should've done a better job describing this error 😅 

This is essentially a programmer error, but still nice to give a better description of the problem.